### ISSUE TITLE
Error refreshing authentication token

### ISSUE THREAD

**Issue by blazer0x** [yadayada/acd_cli#247]
I've been using acd_cli for a few weeks now. I am still in the process of uploading a large amount of data. acd_cli performed flawlessly for over a week of constant uploads without interruption. A couple of days ago though it started throwing "Error refreshing authentication token" errors, and I would have to fetch a new oauth_data token to continue the upload or any other thing like sync.  

I can see that the default "expires_in" time is 3600 (seconds). But previously acd_cli was properly using the refresh_token to refresh the oauth data, and for some reason it has stopped doing this.

Is there a way I can troubleshoot this to determine if the issue is on the Amazon API side, or a local issue with either python or the code itself?

Here is an example of a failure once the token expires:
```
$ acd_cli upload -x 4 /some/path/ /BACKUP/
16-01-11 11:42:15.474 [ERROR] [acd_cli] - Uploading "somefile.ext" failed. RequestError: 400, {"message":"Cannot
complete upload"}.
16-01-11 11:42:16.320 [CRITICAL] [acdcli.api.oauth] - Error refreshing authentication token.
```

After an error such as above, I have to manually update the oauth_data file and then I can upload data for another hour.

$ acd_cli version
acd_cli 0.3.1, api 0.8.8

**Comment by yadayada** [comment]
Judging from the code in acdcli/api/oauth.py, that's a network connection error. It would print Amazon's error message otherwise (if any).

**Comment by blazer0x** [comment]
I don't see how it can be network related. The network connection is stable, and although acd_cli worked flawlessly for days on end, now cannot run for more than an hour without failing to refresh the auth_token. I guess I can try attatching to the process with strace near when it times out and see if I can see anything that is not being reported to stdout.

**Comment by blazer0x** [comment]
I just noticed I also get these errors back to back:
16-01-11 13:49:58.347 [CRITICAL] [acdcli.api.oauth] - Error refreshing authentication token.
16-01-11 13:49:58.348 [ERROR] [acd_cli] - Uploading "file.ext failed. RequestError: 1000, [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:600).

**Comment by Neodark7** [comment]
there is an error when refreshing your auth token, so refresh your auth token

https://tensile-runway-92512.appspot.com/

works for me :)

**Comment by blazer0x** [comment]
Yes, I know how to refresh the auth token. However, I have no desire to do so every hour on the dot for the next several days :P

**Comment by yadayada** [comment]
Please try to connect to appspot
```
echo | openssl s_client -connect appspot.com:443 
```

and to the endpoints (look them up in `endpoint_data` in your cache path.)

**Comment by blazer0x** [comment]
doing that test against appspot.com works, as to the 2 hosts in my endpoint_data file (content-na.drive.amazonaws.com and cdws.us-east-1.amazonaws.com).

Here is some info I found in the acd_cli.log:
16-01-12 12:17:13.042 [INFO] [acdcli.api.oauth] - Token expired at 2016-01-12 10:13:30.322666.
16-01-12 12:17:13.043 [INFO] [acdcli.api.oauth] - Refreshing authentication token.
16-01-12 12:17:13.048 [INFO] [requests.packages.urllib3.connectionpool] - Starting new HTTPS connection (1): tensile-runway-92512.appspot.com
16-01-12 12:17:13.127 [CRITICAL] [acdcli.api.oauth] - Error refreshing authentication token.
16-01-12 12:17:13.128 [ERROR] [acd_cli] - Uploading "somefile.ext" failed. RequestError: 1000, [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:600).
16-01-12 12:17:13.131 [INFO] [acd_cli] - Uploading /path/to/somefile.ext
16-01-12 12:17:13.131 [WARNING] [acdcli.api.backoff_req] - Waiting 112.954014s because of error(s).

**Comment by yadayada** [comment]
>doing that test against appspot.com works

What "Verify return code" do you get?

**Comment by blazer0x** [comment]
```  Verify return code: 0 (ok)

**Comment by yadayada** [comment]
Sorry, I forgot that requests has its own CA bundle. Please try again with your requests path as CApath argument like
```
echo | openssl s_client -CApath `pip3 show requests | grep Location | awk '{print $2}' ` -connect appspot.com:443      
```

**Comment by blazer0x** [comment]
Ah now looks like we got something:
 echo | openssl s_client -CApath `pip3 show requests | grep Location | awk '{print $2}' ` -connect appspot.com:443
CONNECTED(00000003)
depth=2 C = US, O = GeoTrust Inc., CN = GeoTrust Global CA
verify error:num=20:unable to get local issuer certificate

.... <snip>

    Verify return code: 20 (unable to get local issuer certificate)
---
DONE

**Comment by yadayada** [comment]
Maybe installing certifi via pip will help. If not, I suggest you get your own security profile

**Comment by blazer0x** [comment]
I installed certifi, the test with openssl still failed. Trying now with a whitelisted security profile. Thank you for your patience and suggestions.

**Comment by blazer0x** [comment]
Using a security profile didn't help. Still timed out after one hour and error refreshing authentication token. I previously also tried totally uninstalling acd_cli (acd_cli delete-everything ; pip3 uninstall acdcli) and then re-installing again.

I ran strace on the openssl test that fails, and noticed this:

24350 stat("/usr/lib/python3.4/site-packages/578d5c04.0", 0x7ffd627e3180) = -1 ENOENT (No such file or directory)
...
24350 write(1, "---\nNo client certificate CA names sent\n", 40) = 40
...
24350 write(2, "verify error:num=20:unable to get local issuer certificate\n", 59) = 59

Seems odd that it's looking for a file called "578d5c04.0". Just for fun I tried creating a symlink in the site-packages directory with the name "578d5c04.0" linked to  requests/cacert.pem:

lrwxrwxrwx 1 root root 19 Jan 12 17:03 578d5c04.0 -> requests/cacert.pem

The openssl test now passes! I am doing another upload test, It will take me an hour to find out if this crazy fix worked :)

**Comment by blazer0x** [comment]
Alas, despite now passing the openssl test, the token refresh and SSL: CERTIFICATE_VERIFY_FAILED error is still happening. It is even more of a pain now, as since I am using a security profile, when the auth token expires, I have to re-auth the security profile page as well as the page that updates oauth_data.

**Comment by yadayada** [comment]
>I have to re-auth the security profile page **as well as the page that updates oauth_data.**

I think you're doing it wrong.

**Comment by blazer0x** [comment]
Yeah I don't know why, but deven though I am using a security profile, it is also still using the oath_data token. When the token expires I have to delete oauth_data, and then when I do a sync or upload it sends me through the auth of both the security_profile page (updates client_data) and the oauth from tensile-runway.  I'm at a loss of what is wrong, I keep just restarting it every hour that I can, and losing 8-12 hours per day of upload time while I am asleep etc.

**Comment by yadayada** [comment]
Don't delete `oauth_data`, the file is also needed if you have a security profile.
For testing purposes, you can manually "expire" your tokens by increasing the expiration time by 3600 seconds or more in your oauth_data.

Please check that openssl works with https://api.amazon.com.

If you can't fix the openssl/requests issue, you should manually edit acdcli/api/oauth.py and modify the relevant requests calls so as to not verify SSL certificates by adding `verify=False` to the parameter list.

**Comment by blazer0x** [comment]
The verify trick did not work. However I DID get it working, by updating my root cert bundle. I did this via "curl http://curl.haxx.se/ca/cacert.pem >> /tmp/cacert.pem /usr/lib/python3.4/site-packages/certifi/cacert.pem"  (the curl site maintains a bundle of up-to-date root certs). Immediately after doing this acd_cli was working for me again, and it automatically updated the oauth_data file by itself as it should.

Long story short, I do not believe this was a bug with acd_cli, but rather some weirdness with my servers root certificates (Centos 6.7). Once the certs were updated, everything is fine.
